packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,601 @@
+    }
+
+    void recursiveExportSymbol(core::GlobalState &gs, bool firstSymbol, core::ClassOrModuleRef klass) {
+        // There is a loop between a class and its singleton, and this is the easiest way to detect it.
+        if (!firstSymbol && klass.data(gs)->flags.isExported) {
+            return;
+        }
+
+        // We only mark symbols from this package.
+        if (!this->definedByThisPackage(gs, klass)) {
+            return;
+        }
+
+        klass.data(gs)->flags.isExported = true;
+
+        for (const auto &child : klass.data(gs)->members()) {
+            if (child.second.isClassOrModule()) {
+                recursiveExportSymbol(gs, false, child.second.asClassOrModuleRef());
+            } else if (child.second.isFieldOrStaticField()) {
+                child.second.asFieldRef().data(gs)->flags.isExported = true;
+            }
+        }
+    }
+
+    void exportParentNamespace(core::GlobalState &gs, core::ClassOrModuleRef owner) {
+        while (owner.exists() && !owner.data(gs)->flags.isExported && this->definedByThisPackage(gs, owner)) {
+            owner.data(gs)->flags.isExported = true;
+            owner = owner.data(gs)->owner;
+        }
+    }
+
+    // Lookup the package name on the given root symbol, and mark the final symbol as exported.
+    void exportRoot(core::GlobalState &gs, core::ClassOrModuleRef sym) {
+        // For a package named `A::B`, the ClassDef that we see in this pass is for a symbol named
+        // `<PackageSpecRegistry>::A::B`. In order to make the name `A::B` visibile to packages that have imported
+        // `A::B`, we explicitly lookup and export them here. This is a design decision inherited from the previous
+        // packages implementation, and we could remove it after migrating Stripe's codebase to not depend on package
+        // names being exported by default.
+        for (auto name : this->package.fullName()) {
+            auto next = sym.data(gs)->findMember(gs, name);
+            if (!next.exists() || !next.isClassOrModule()) {
+                sym = core::Symbols::noClassOrModule();
+                return;
+            }
+
+            sym = next.asClassOrModuleRef();
+        }
+
+        sym.data(gs)->flags.isExported = true;
+    }
+
+    // Mark both `::A::B` and `::Test::A::B` as exported (given package `A::B`).
+    // This will terminate the recursive exporting in `exportParentNamespace`, as it stops when it
+    // hits either the root or an exported symbol.
+    void exportPackageRoots(core::GlobalState &gs) {
+        this->exportRoot(gs, core::Symbols::root());
+
+        auto test = core::Symbols::root().data(gs)->findMember(gs, core::Names::Constants::Test());
+        if (test.exists() && test.isClassOrModule()) {
+            this->exportRoot(gs, test.asClassOrModuleRef());
+        }
+    }
+
+    void setPackageLocs(core::MutableContext ctx, core::LocOffsets loc, core::ClassOrModuleRef sym) {
```

Comment:
Not sure I fully get what's going on here. Can you document?

Reply:
Yep, this definitely warrants some explanation. I'll add a comment.